### PR TITLE
API documentation

### PR DESCRIPTION
**Pull request by snej**
Adds detailed API documentation for the major interfaces and some of the other types. This is based on my experience with the Walrus implementation and may not be 100% correct.

I've left `TODO` comments indicating things I'm unsure of, and some other things that look like they could be removed.

No code changes.

### CODE REVIEW

**Comment by snej** [comment]
Note: I removed the constants FeedOpCheckpointStart and FeedOpCheckpointEnd because they seem obsolete: there are no references to them anywhere in the SG codebase.

File: bucket.go [hunk 1]
```diff
@@ -52,36 +54,54 @@ const (
...
 // MutationFeedStore is a store that supports a DCP or TAP streaming mutation feed.
 type MutationFeedStore interface {
+	// The number of vbuckets of this store; usually 1024.
 	GetMaxVbno() (uint16, error)
+
+	// Starts a new DCP event feed. Events will be passed to the callback function.
+	// To close the feed, pass a channel in args.Terminator and close that channel.
+	// - args: Configures what events will be sent.
+	// - callback: The function to be called for each event.
+	// - dbStats: TODO: What does this do? Walrus ignores it.
```
Comment (adamcfraser) [inline bucket.go:90]: Looks like this is solely used as a mechanism to pass through rollback counts as a stat at the moment, and walrus doesn't rollback.

File: bucket.go [hunk 2]
```diff
@@ -109,75 +131,294 @@ type MutateInOptions struct {
...
+	// - cas: Expected CAS value
+	// - v: The value to set. Will be marshaled to JSON unless it is a `[]byte` or `*[]byte`
+	// - opt: Options; see WriteOptions for details
+	// Return values:
+	// - casOut: The new CAS value
+	// - err: Error, if any. May be CasMismatchErr
 	WriteCas(k string, flags int, exp uint32, cas uint64, v interface{}, opt WriteOptions) (casOut uint64, err error)
+
+	// Deletes a document by setting its value to nil, making it a tombstone.
+	// System xattrs are preserved but user xattrs are removed.
```
Comment (adamcfraser) [inline bucket.go:221]: This feels like partial coverage of the xattr behaviour (doesn't describe that they are unchanged on WriteCas, deleted on Add, etc).  Could either add a note on each KVStore operation, or document elsewhere.

File: bucket.go (hunk 1 above)
Comment (snej) [reply bucket.go]: Thanks. I've added that to the comments (in both methods where dbStats appears.)

File: bucket.go (hunk 2 above)
Comment (snej) [reply bucket.go:223]: That's good to know -- I wasn't aware of the behavior of Add, for example. I've added notes on both those methods. I'd like to add xattr notes on the other methods, but someone will need to tell me what the effects are.

**Comment by torcolvin** [comment]
These changes were merged with a separate PR.

**Pull request closed without merging**